src/zfs-util.c

```diff
@@ -9,29 +9,75 @@
 // Buffer for reading ZFS command output
 #define BUFSIZE 16
 
-int execute(char *command, char needOutput, char **output, char *param[]) {
-	int pip[2];
-	pid_t pid;
-	int status;
+int execute(char *command, char needOutput, char **output, char *param[], char needPassword, char *prompt) {
+	int pip[2], pwpip[2];
+	pid_t pid, pwpid;
+	int status, pwstatus;
 	char *linebuffer;
 	size_t size;
 	int nRead;
+	char **cmdline;
+
+	if (needPassword == 1) {
+		// Execute password prompt
+		if (pipe(pwpip) == -1) {
+			perror("Cannot create password pipe");
+			exit(1);
+		}
+		pwpid = fork();
+		if (pwpid == 0) {
+			// Redirect stdout to pipe
+			while ((dup2(pwpip[1], STDOUT_FILENO) == -1) && (errno == EINTR)) {}
+			if (errno != 0) {
+				perror("Cannot connect stdout to parent password pipe");
+				exit(1);
+			}
+			// Close pipe, we won't need it anymore
+			close(pwpip[1]);
+			// Close stdin to prevent weird redirection
+			close(0);
+			// Execute
+			if (prompt == NULL) {
+				cmdline = (char*[]) { SYSTEMD_ASK_PASS_CMD, NULL };
+			} else {
+				cmdline = (char*[]) { SYSTEMD_ASK_PASS_CMD, prompt, NULL };
+			}
+			execv(SYSTEMD_ASK_PASS_EXE, cmdline);
+			exit(254);
+		} else if (pwpid < 0) {
+			perror("Cannot fork");
+			close(pwpip[0]);
+			close(pwpip[1]);
+			return pwpid;
+		}
+	}
 
 	// Execute
 	if (needOutput == 1) {
 		if (pipe(pip) == -1) {
-			perror("Cannot create pipe\n");
+			perror("Cannot create output pipe");
```

Comment:
Where did my \n go?

Reply:
The removal of the newlines came from a discovery that perror's string is gonna get `: <some error>` appended to it when it's triggered, which made the newlines seem wrong, so I dropped them. If this was for some reason intentional, I can revert this.

Reply:
Never mind - thank you for the correction ;)